### ISSUE TITLE
Automatische SystemSoftware-Updates

### ISSUE THREAD

**Issue by patrickhilker** [mittwald/feature-requests#70]
**Welches Problem möchtest du lösen? Wann tritt es auf?**
Als Agentur/Entwickler möchte ich jederzeit die aktuellste Software für den Betrieb meiner Apps nutzen. Ich erwarte von meinem Webhoster, dass die von mir eingesetzte Software jederzeit sicher ist und keine Sicherheitslücken enthält. (Gemeint ist hier PHP, ImageMagick - und explizit **nicht Apps** wie WordPress, TYPO3, ...)

**Welche Lösungsideen hast du?**
mittwald sollte automatisch Updates, die kompatibel zu den von mir installierten Apps sind, für meine Software installieren. Dabei möchte ich allerdings konfigurieren können, für welche Software das gilt, und ob ich dabei auch mögliche Breaking Changes akzeptiere.

**Hast du zusätzliche Informationen (wie z.B. Screenshots)?**
Am wichtigsten sind Updates vermutlich für die PHP-Version.

**Comment by holmey** [comment]
Ja PHP ist wichtig, aber auch die Datenbank. Wahrscheinlich wird es schwer hier automatisiert zu patchen. Wie könnte ihr dann garantieren, dass die Anwendungen weiter ohne Einschränkungen laufen.
Eine andere Idee wäre es, einen Benachrichtigungs-Agenten/Monitoring Service zur Verfügung zu stellen. Dieser könnte dann diverse Health-Checks durchführen. So etwas wie https://ohdear.app/, bzw. hier insbesondere das feature Health Checks.
Eine einfache Komma getrennte Liste für E-Mails die benachrichtigt werden sollten. Dann können die Entwickler nach Eingang einer Warnung handeln.

Übergreifende checks:
- Speicherplatz geht aus
- ...

Default Checks auf App Ebene z.B. für die PHP App:
- CVE für PHP Version oder DB
- Limits für die Performance
- ...

ggf. eine eigene API, über die man via CronJob eigene Tests durchführen kann und die Message queue füllt (siehe z.B. https://ohdear.app/docs/integrations/the-oh-dear-api ):
- CMS Projekt Verzeichnis-Berechtigungen falsch gesetzt
- Kritische Plugin-Updates
- CVE für Software erkannt (composer audit)
- PHP Version stimmt nicht mit Plattform in composer überein
- ...

So ein feature würde gleichzeitig vielen Agenturen helfen ihren Sorgfaltspflichten, oder verhandelten Leistungsbeschreibungen nachzukommen.

**Comment by patrickhilker** [comment]
> Wie könnte ihr dann garantieren, dass die Anwendungen weiter ohne Einschränkungen laufen.

Wir haben einige Ideen dazu, die eine automatische Prüfung inkl. Rollback enthalten, aber du kannst dir sicher vorstellen, dass das alles andere als trivial ist.

Ich finde einige spannende Punkte, wie die automatischen Benachrichtigungen, die ich mir gerne mal mitnehme.

Unabhängig davon: Über unsere API kannst du ja jederzeit die installierte Software und die von uns bereitgestellten Versionen auslesen, abgleichen und auch updaten. Theoretisch könnte man sich das also schon jetzt selbst basteln.

**Comment by holmey** [comment]
Mir würde es auch gar nicht darum gehen automatisiert größere Upgrades durchzuführen. Ich habe eher ein Feature im Sinn, welches [UnattendedUpgrades](https://wiki.debian.org/UnattendedUpgrades) entspricht.

Wie macht ihr das eigentlich bei Applikationen im alten Kundencenter. Wird dort z.B. PHP, Apache usw. auf Patch-Level aktualisiert?

**Comment by patrickhilker** [comment]
> Mir würde es auch gar nicht darum gehen automatisiert größere Upgrades durchzuführen. Ich habe eher ein Feature im Sinn, welches [UnattendedUpgrades](https://wiki.debian.org/UnattendedUpgrades) entspricht.

Magst du ein bisschen genauer erklären, was du da erwarten würdest?

> Wie macht ihr das eigentlich bei Applikationen im alten Kundencenter. Wird dort z.B. PHP, Apache usw. auf Patch-Level aktualisiert?

Der Apache und die gesamte Software auf dem Server, die nicht durch den Kunden verwaltet wird, wird von unseren Administratoren aktuell gehalten. Die meisten Softwareversionen, die man selbst im Kundencenter verwalten kann, werden von uns nur in Notfällen angefasst. Bei einigen Softwares gibt es allerdings die Möglichkeit, die Version auf "xxx-latest" zu setzen (bspw. `PHP 8.3-latest`), muss aber gestehen, dass ich nicht 100 % weiß, wie sich das dann genau verhält (bin quasi nur im mStudio unterwegs).

Da möchten wir im mStudio definitiv besser und nachvollziehbarer werden. :-)

**Comment by holmey** [comment]
Es geht mir lediglich um kritische Sicherheitsupdates, die auf irgendeine Art sicher automatisiert installiert werden können durch euch. Wenn wir also eine PHP App anlegen würde ich mir wünschen, dass z.B. sobald ein Patch für ein CVE vorhanden ist, dieser automatisch installiert wird. Was ich nicht meine ist, dass automatisch von sagen wir PHP 8.2. auf 8.3 aktualisiert wird.

**Comment by patrickhilker** [comment]
Ja, die Möglichkeit haben wir und würden das definitiv auch jetzt schon machen, wenn es sicherheitskritisch ist, sowohl im Bestandssystem als auch im neuen System. :-)

**Comment by tehplague** [comment]
Ich denke, innerhalb einer Minor-Version wären automatische Patch-Updates auch jetzt eine sehr sinnvolle Ergänzung. Denn wenn ich z.B. aktuell eine laufende App auf PHP 8.4.1 habe, kann ich auch jetzt in mStudio nur noch 8.4.3 auswählen. Allerdings muss ich das in allen bestehenden Apps auch manuell tun oder per `mw` skripten.
Das widerspricht teils meinem Verständnis eines Managed-Hostings, bei dem ich zumindest gerne einen zuschaltbaren Automatismus hätte. Denn so wie es aktuell ist, skaliert es für Agenturen ohne Zutun per Skript überhaupt nicht. Letztlich ist es ja auch in eurem Interesse, dass Update-Rückstände so abgebaut werden, anstatt, dass Apps mit offenen CVEs im PHP-Interpreter herumliegen, weil niemand regelmäßig durch alle Projekte stiefelt und die Versionsstände auf Updates prüft.

Projekte wie https://github.com/thueske/mstudio-php-viewer sind zwar nett, aber kein Ersatz für automatische Patch-Updates. Das Thema ist auch nicht auf PHP beschränkt, sondern gleichermaßen für Ghostscript, libpoppler oder die Datenbanken.

**Comment by patrickhilker** [comment]
PHP wird jetzt automatisch aktuell gehalten (neue Patch-Level werden automatisch aktualisiert).
